File: app/lib/features/tracking/domain/entities/entry.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'entry.freezed.dart';
part 'entry.g.dart';

enum EntryType {
  feeding,
  diaper,
  sleep,
}

enum FeedingType {
  breastLeft,
  breastRight,
  bottle,
  formula,
}

enum DiaperType {
  wet,
  dirty,
  both,
}

enum SleepQuality {
  good,
  fair,
  poor,
}

@freezed
class Entry with _$Entry {
  const factory Entry({
    required String id,
    required String familyId,
    required String childId,
    required EntryType type,
    required DateTime timestamp,
    required DateTime createdAt,
    required String createdBy,
    required DateTime updatedAt,
    required String updatedBy,
    String? notes,
    @Default([]) List<String> mediaIds,

    // Feeding-specific
    FeedingType? feedingType,
    double? amount,
    int? duration,

    // Diaper-specific
    DiaperType? diaperType,

    // Sleep-specific
    DateTime? endTime,
    SleepQuality? quality,
  }) = _Entry;

  factory Entry.fromJson(Map<String, dynamic> json) => _$EntryFromJson(json);
}

extension EntryExtensions on Entry {
  bool get isFeeding => type == EntryType.feeding;
  bool get isDiaper => type == EntryType.diaper;
  bool get isSleep => type == EntryType.sleep;

  Duration? get sleepDuration {
    if (!isSleep || endTime == null) return null;
    return endTime!.difference(timestamp);
  }
}

File: backend/internal/domain/entry.go
package domain

import "time"

type EntryType string

const (
	EntryTypeFeeding EntryType = "feeding"
	EntryTypeDiaper  EntryType = "diaper"
	EntryTypeSleep   EntryType = "sleep"
)

type Entry struct {
	ID        string    `json:"id" firestore:"-"`
	FamilyID  string    `json:"familyId" firestore:"-"`  // Parent reference
	ChildID   string    `json:"childId" firestore:"-"`   // Parent reference
	Type      EntryType `json:"type" firestore:"type"`
	Timestamp time.Time `json:"timestamp" firestore:"timestamp"`
	CreatedAt time.Time `json:"createdAt" firestore:"createdAt"`
	CreatedBy string    `json:"createdBy" firestore:"createdBy"`
	UpdatedAt time.Time `json:"updatedAt" firestore:"updatedAt"`
	UpdatedBy string    `json:"updatedBy" firestore:"updatedBy"`
	Notes     string    `json:"notes,omitempty" firestore:"notes,omitempty"`
	MediaIDs  []string  `json:"mediaIds,omitempty" firestore:"mediaIds,omitempty"`

	// Feeding-specific fields
	FeedingType FeedingType `json:"feedingType,omitempty" firestore:"feedingType,omitempty"`
	Amount      *float64    `json:"amount,omitempty" firestore:"amount,omitempty"`   // ml or oz
	Duration    *int        `json:"duration,omitempty" firestore:"duration,omitempty"` // minutes

	// Diaper-specific fields
	DiaperType DiaperType `json:"diaperType,omitempty" firestore:"diaperType,omitempty"`

	// Sleep-specific fields
	EndTime *time.Time   `json:"endTime,omitempty" firestore:"endTime,omitempty"`
	Quality SleepQuality `json:"quality,omitempty" firestore:"quality,omitempty"`
}

type FeedingType string

const (
	FeedingTypeBreastLeft  FeedingType = "breast_left"
	FeedingTypeBreastRight FeedingType = "breast_right"
	FeedingTypeBottle      FeedingType = "bottle"
	FeedingTypeFormula     FeedingType = "formula"
)

type DiaperType string

const (
	DiaperTypeWet   DiaperType = "wet"
	DiaperTypeDirty DiaperType = "dirty"
	DiaperTypeBoth  DiaperType = "both"
)

type SleepQuality string

const (
	SleepQualityGood SleepQuality = "good"
	SleepQualityFair SleepQuality = "fair"
	SleepQualityPoor SleepQuality = "poor"
)
